### ISSUE TITLE
perf(compute): qualify exact FP32 tiled Q4_K prefill on gfx1151

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12731]
# perf(compute): qualify exact FP32 tiled Q4_K prefill on gfx1151

```routing
lane: internal/compute
paths: ["internal/compute/build_vulkan.ps1", "internal/compute/shaders/q4k_matmul_tiled_fp32.comp", "internal/compute/vulkan.go", "internal/compute/vulkan_backend.h", "internal/compute/vulkan_shim.cpp", "internal/compute/vulkan_q4k_tiled_fp32_test.go", "internal/computebuild/vulkan.go", "internal/computebuild/computebuild_test.go"]
expected_steps: 8
repo: fak
```

## Parent context

Public issue #12177 owns the cooperative-matrix/WMMA Q4_K prefill experiment and its existing `>=300 tok/s` physical floor. That issue remains open. This leaf tracks the distinct correctness-preserving fallback discovered while diagnosing it: a shared-LDS, exact-FP32 tiled path with no cooperative-matrix instructions.

## Current state

The native multi-token Q4_K path computes each token/output pair independently. The #12177 candidate shares decoded values through LDS and uses F16 cooperative matrices, but its accumulation differs from the scalar FP32 order. On the target device, actual-width diagnostics at `K=17408` failed the unchanged normalized-relative-L2 `<=0.0001` gate for both the three-product and four-product cooperative variants. The four-product correction changed residual only marginally and did not clear the gate.

A bounded diagnostic instead stages activation and Q4_K weight tiles in LDS and accumulates FP32 in the scalar K order. Its SPIR-V contains zero cooperative-matrix operations. On the same actual-width fixtures, its outputs were bit-identical to the forced GPU scalar path at `K=5120` and `K=17408`. A P32 whole-model smoke and split-prefill continuity check also produced exact compared logits and token IDs.

The same diagnostic then passed six alternating P128 pairs after an excluded warmup. Using type-7 quantiles and exhaustive paired bootstrap over all `6^6` resamples, scalar/tiled P50 was `9.5124884945/5.205367918 s`, P90 was `9.516493978/5.2074679645 s`, and the ratio-of-medians speedup was `1.8274382607` with 95% interval `[1.8249559143, 1.8333966901]`. This is a point-in-time diagnostic result. Its archive contains the test harness plus a substituted shader bundle rather than the final production source, and its receipt lacks clocks, power, thermal, and memory samples. Production-source qualification and the default gate remain open.

## Why this is next

The exact tiled path keeps the useful cross-token data reuse from the larger prefill design while removing the error source that blocks whole-model admission. It can ship independently from WMMA and can be evaluated against the current scalar path with an exact-output oracle and a matched physical timing distribution.

## Value

- Centrality: Core
- P1: advanced - removes repeated Q4_K decode work across prompt tokens.
- P2: preserved - retains resident buffers, exact FP32 accumulation order, and the scalar rollback path.
- P3: advanced - reuses the existing native Vulkan sequence and source-bound model witness.
- P4: preserved - confines production changes to compute and shader registration without changing an exported API.

## Core through-line

Resident F32 activations plus Q4_K weights -> capability-gated 2D shared-LDS tile -> scalar-order FP32 accumulation -> resident output -> exact scalar parity and route attribution -> repeated physical KEEP/REJECT receipt -> narrow gfx1151 default only after the gate passes.

## Working spine

Add `q4k_matmul_tiled_fp32.comp`, register the SPIR-V in the canonical builder, expose a dedicated availability flag and dispatch counter, and route eligible multi-token Q4_K calls through the new pipeline. `FAK_VULKAN_Q4K_TILED_FP32=candidate` explicitly selects it; a non-candidate nonempty value disables it. `FAK_VULKAN_Q4K_ARM=scalar` or `0` always wins.

## Gold-plating boundary

- No cooperative-matrix, WMMA, tensor-core, or F16 accumulation claim.
- No change to Q4_K encoding, model architecture, attention, decode kernels, or exported API.
- No device-wide default. Automatic admission is limited to the exact qualified AMD vendor/device pair and `P>=32`.
- No inherited `>=300 tok/s` requirement from #12177. This leaf requires a repeated matched net improvement with unchanged outputs.
- No performance claim from the single-sample correctness smoke.

## Verifiable Witness

```bash
pwsh internal/compute/build_vulkan.ps1 build
go test -count=1 ./internal/computebuild -run '^(TestVulkanShadersCompleteness|TestVerifyVulkanBinaryReceiptIdentity.*)$'
go test -tags vulkan -count=1 ./internal/compute -run '^TestVulkanQ4KTiledF32'
```

## Witness

The commands above bind shader compilation, bundle completeness, exact actual-width primitive parity, and route attribution. The existing model-level sequence fixture supplies the separate source-bound hardware witness; production scope remains the two routed packages.

### File:Line Seam

- `internal/compute/shaders/q4k_matmul_tiled_fp32.comp:1` — new exact-FP32 tiled kernel.
- `internal/compute/vulkan_shim.cpp` (`fvk_q4k_matmul_f32`) — Q4_K pipeline admission and 2D dispatch.
- `internal/compute/vulkan.go` (`VulkanQ4KTiledF32Available`) — public-internal capability and route counter surface.
- `internal/computebuild/vulkan.go` (`VulkanShaders`) — canonical shader bundle registration.

### Blast Radius & Affected Lanes

- **Primary lane:** `internal/compute`; build registration in `internal/computebuild`.
- **Test-only witness:** the existing model-level sequence prefill fixture.
- **Unaffected lanes:** other backends, quantization formats, model scheduling, private platform, and exported SDKs.

### Quarantined Fallback Mechanism

The current scalar Q4_K shader remains the fallback. Missing pipeline support, unsupported device identity, short prompts, explicit disable, and `FAK_VULKAN_Q4K_ARM=scalar` or `0` all select the scalar path. The new path stays explicit-only until the physical distribution and exact-output gates pass.

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] The exact-FP32 tiled shader compiles, validates, and is registered in the canonical Vulkan bundle; SPIR-V inspection records zero cooperative-matrix instructions.
- [ ] 2. [SW-VERIFIED] P32 actual-width fixtures at `K=5120` and `K=17408`, including signed cancellation, are bit-identical to forced GPU scalar and pass the unchanged CPU oracle (`relative L2 <=0.0001`, cosine `>=0.99999`, exact argmax).
- [ ] 3. [SW-VERIFIED] Selector tests prove explicit candidate admission for `P>1`, explicit disable, P1/P16 scalar behavior, and precedence for `FAK_VULKAN_Q4K_ARM=scalar` and `0`.
- [ ] 4. [HW-WITNESSED] Source-bound P32 whole-model and split-prefill continuity witnesses preserve exact compared logits and generated token IDs, use the native sequence route without fallback, and record positive tiled dispatch only during prefill.
- [ ] 5. [HW-WITNESSED] After an excluded warmup, at least six alternating scalar/candidate P128 pairs report raw samples, P50/P90, a paired uncertainty interval, clocks, power, thermals, memory state, dispatch counts, and every available route byte counter; unavailable transfer counters are named explicitly. KEEP requires candidate P50 faster than scalar with the paired 95% interval wholly above a 1.0 speedup and no correctness change.
- [ ] 6. [SW-VERIFIED] Automatic admission, if criterion 5 passes, is limited to AMD vendor `0x1002`, device `0x1586`, and `P>=32`; unset on all other devices and shapes remains scalar.
- [ ] 7. [SW-VERIFIED] Affected compute/build tests and boundary/leak checks pass on the committed source.

## Done condition

The production source is landed with exact-output tests and scalar fallback, then a source-bound physical receipt records a repeated net prefill improvement. Default admission remains narrow and requires every hardware and selector criterion above. #12177 remains open for the separate cooperative-matrix mechanism and its `>=300 tok/s` floor.

## Acceptance gate

All software criteria and the committed-source physical correctness/performance criteria must pass. Any failure retains the scalar default and leaves this issue open.

## Lane

`internal/compute`; two production packages plus one model-only hardware witness.

## Closure binding

The implementation commit references this issue and #12177 with `(fak compute)`. Close only after the committed-source physical KEEP receipt; a diagnostic shader, primitive pass, or single sample is insufficient.

## Work estimate
Estimate: 5 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 5/100 points.

## Completion standard
production

## Target operating envelope
- paired P128 median speedup: >= 1.0 ratio
- exact output match rate: = 100 percent

## Witnessed operating envelope
- paired P128 median speedup: 1.8274382607 ratio
- exact output match rate: 100 percent

**Comment by anthony-chaudhary** [comment]
Recovery/integration update (2026-09-12): the original checkpoint commit is unavailable from local objects and origin, but the full archived source was recovered and its 200,429,082 bytes match SHA-256 `3136275eba03235090aeecabf8a04338656ed835e8ae9f517bd749d1564ff086`. Reconstructing the historical base plus exactly the ten candidate paths yields the original tree `deb2251639b60420dab1ee86cc7a3d5f744a3c04`.

Current integration base is `f1bcc34e0`. Commit `f20977cbdeec0c18b4a332d86396c177cea8e71d` changed the forced scalar Q4_K route to eight accumulation chains. The recovered tiled candidate therefore cannot inherit its historical exact-parity or performance acceptance. An isolated candidate now follows the current scalar chains and final fold; the strict bitwise, CPU-oracle, and model tests remain unchanged. Its shader compiles, and independent source/SPIR-V review found no concrete arithmetic/indexing defect. These are software review results only.

The merged 47-module registry passes `go test -count=1 ./internal/computebuild -run '^(TestVulkanShadersCompleteness|TestVerifyVulkanBinaryReceiptIdentity.*)$'` in WSL. Shared shim integration is queued behind existing issue ownership. Numerical execution, whole-model/continuation, fresh physical paired timing, and landing are pending. No default or performance qualification is claimed. The historical P128 receipt does not establish cold unique-token ingest performance at 5k, 10k, or 25k. This issue remains open.

**Comment by anthony-chaudhary** [comment]
Software checkpoint update: the candidate remains explicit opt-in (`FAK_VULKAN_Q4K_TILED_FP32=candidate`); unset/empty/default select scalar, and scalar override precedence is preserved. The full-model Q4_K witness now pins and reports `FAK_VULKAN_GDN_Q8_PANEL=0` to avoid mixing independent optimizations.

Actual llvmpipe software Vulkan execution passed both `TestVulkanQ4KTiledF32PrefillMatchesCPUReference` and `TestVulkanQ4KTiledF32DefaultRemainsScalar`, including every subcase, without skips. At K=5120 and K=17408, candidate/scalar bit mismatches were zero. CPU residuals were `1.4759746e-7` and `8.5934533e-5`, below the unchanged `1e-4` limit. Replacing only the tiled shader with the recovered historical single-chain shader in a separate diagnostic bundle makes the same exact tests fail: 3,943 and 4,128 mismatches respectively. This is software fail-before/pass-after evidence, not physical GPU or performance qualification.

Tagged compute compilation, direct C++ shim compilation, full computebuild tests, public staged leak audit and five-gate boundary audit pass. The complete shader sweep is not green: the pre-existing cooperative shader crashes the installed WSL compiler, recorded on #12177. Model compilation and governed final validation/landing remain pending. No Halo qualification, automatic default, deployment, long-context parity claim, or issue closure is authorized by these results.

**Comment by anthony-chaudhary** [comment]
Final validation hold for the current opt-in checkpoint: the normal exact-path `fak validate --mine ... --smoke --timeout 4m --json` invocation returned `WSL_CAPABILITY_PREFLIGHT_FAILED`. Its WSL preflight took 14,543 ms and exceeded the fixed 14.5-second allowance before allocating a validation workspace. Build, vet, test and smoke phases were skipped; this is not a passing validation or evidence of a source failure. No fallback or bypass was selected, and no landing/push was attempted.

The next checkable step is to restore responsive WSL capability preflight and rerun the exact ten-path validation. The independent whole-place/controller authority and full shader compiler gates remain separate pending requirements. Source and fail-before/pass-after software evidence are preserved; this issue stays open.
